Core/include/Acts/TrackFitting/detail/GsfActor.hpp

```diff
@@ -0,0 +1,728 @@
+
+    // Create all possible new components
+    for (const auto& gaussian : mixture) {
+      // Here we combine the new child weight with the parent weight.
+      // However, this must be later re-adjusted
+      const auto new_weight = gaussian.weight * old_weight;
+
+      if (new_weight < m_cfg.weightCutoff) {
+        ACTS_VERBOSE("Skip component with weight " << new_weight);
+        continue;
+      }
+
+      if (gaussian.mean < 1.e-8) {
+        ACTS_WARNING("Skip component with gaussian " << gaussian.mean << " +- "
+                                                     << gaussian.var);
+        continue;
+      }
+
+      // compute delta p from mixture and update parameters
+      auto new_pars = old_bound.parameters();
+
+      const auto delta_p = [&]() {
+        if (state.stepping.navDir == NavigationDirection::forward)
+          return p_prev * (gaussian.mean - 1.);
+        else
+          return p_prev * (1. / gaussian.mean - 1.);
+      }();
+
+      throw_assert(p_prev + delta_p > 0.,
+                   "new momentum after bethe-heitler must be > 0, p_prev= "
+                       << p_prev << ", delta_p=" << delta_p
+                       << ", gaussian mean: " << gaussian.mean);
+      new_pars[eBoundQOverP] = old_bound.charge() / (p_prev + delta_p);
+
+      // compute inverse variance of p from mixture and update covariance
+      auto new_cov = std::move(old_bound.covariance());
+
+      if (new_cov.has_value()) {
+        const auto varInvP = [&]() {
+          if (state.stepping.navDir == NavigationDirection::forward) {
+            const auto f = 1. / (p_prev * gaussian.mean);
+            return f * f * gaussian.var;
+          } else {
+            return gaussian.var / (p_prev * p_prev);
+          }
+        }();
+
+        (*new_cov)(eBoundQOverP, eBoundQOverP) += varInvP;
+        throw_assert(
+            std::isfinite((*new_cov)(eBoundQOverP, eBoundQOverP)),
+            "cov not finite, varInvP=" << varInvP << ", p_prev=" << p_prev
+                                       << ", gaussian.mean=" << gaussian.mean
+                                       << ", gaussian.var=" << gaussian.var);
+      }
+
+      // Set the remaining things and push to vector
+      componentCaches.push_back(
+          {ParameterCache{new_weight, new_pars, new_cov}, metaCache});
+    }
+  }
+
+  template <typename stepper_t>
+  void reduceComponents(const stepper_t& stepper, const Surface& surface,
+                        std::vector<ComponentCache>& cmps) const {
+    // Final component number
+    const auto final_cmp_number = std::min(
+        static_cast<std::size_t>(stepper.maxComponents), m_cfg.maxComponents);
+
+    auto proj = [](auto& a) -> decltype(auto) { return std::get<0>(a); };
+
+    // We must differ between surface types, since there can be different
+    // local coordinates
+    // TODO add other surface types
+    switch (surface.type()) {
+      case Surface::Cylinder: {
```

Comment:
This has been a long-standing problem, for the new geometry, I will add an option to approximate cylinders by multiple surfaces as well ...